New Resource: `azurerm_site_recovery_replication_recovery_plan`;New Data Source: `azurerm_site_recovery_replication_recovery_plan` (#19940)
hashicorp/terraform-provider-azurerm: website/docs/r/site_recovery_replication_recovery_plan.html.markdown

---
subcategory: "Recovery Services"
layout: "azurerm"
page_title: "Azure Resource Manager: azurerm_site_recovery_replication_recovery_plan"
description: |-
    Manages a Site Recovery Replication Recovery Plan on Azure.
---

# azurerm_site_recovery_replication_recovery_plan

> Manages a Azure Site Recovery Plan within a Recovery Services vault. A recovery plan gathers machines into recovery groups for the purpose of failover.

Reviewer: ```suggestion
Manages an Azure Site Recovery Plan within a Recovery Services vault. A recovery plan gathers machines into recovery groups for the purpose of failover.
```

  name                      = "example-recover-plan"
  resource_group_name       = azurerm_resource_group.target.name
  recovery_vault_name       = azurerm_recovery_services_vault.target.name
  source_recovery_fabric_id = azurerm_site_recovery_fabric.source.id
  target_recovery_fabric_id = azurerm_site_recovery_fabric.target.id

  recovery_groups {
    group_type                 = "Boot"
    replicated_protected_items = [azurerm_site_recovery_replicated_vm.test.id]
  }
  recovery_groups {
    group_type = "Failover"
  }
  recovery_groups {
    group_type = "Shutdown"
  }

}
```

## Argument Reference

The following arguments are supported:

* `name` - (Required) The name of the Replication Plan.

* `resource_group_name` - (Required) Name of the resource group where the vault that should be updated is located.

* `recovery_vault_name` - (Required) The name of the vault that should be updated.

* `source_recovery_fabric_id` - (Required) ID of source fabric to be recovered from. Changing this forces a new Replication Plan to be created.

* `target_recovery_fabric_id` - (Required) ID of target fabric to recover. Changing this forces a new Replication Plan to be created.

> * `recovery_group` - (Required) Three or more `recovery_group` block. At least one `recovery_group` for every `group_type` needed.

Reviewer: Not quite understand the second sentence. `group_type` is one element of `recovery_group`, so each `recovery_group` has a `group_type` ?

Author: yes, and there should be at least three `recovery_group` Maybe we can make `recovery_group` a computed one to avoid this?

Reviewer: Can it be `computed`? Isn't `recovery_group` a configurable property?

Author: I remember there is some trouble as it's `typeSet`.. if I set it to `computed` there might be a diff, let me give it a try.

Reviewer: The default hash for `Set` will ignore the `computed` properties. The problem is: if this property is configurable by users, it should not be set as `computed`?

Author: it's configurable and if the users do not configure it, it will also be generated by service. 
I have not set it to `computed` and asking users to set at least one group for every type.
My problem is the block, itself is `computed`, while in this situation, it get response with 3 `recovery_group`, but only one in the config file, then there will be a diff.

Reviewer: Souds like a backend problem. Normally, the backend should not change the configuration sent by the frontend.

---

A `recovery_groups` block supports the following:

*  `group_type` - (Required) The Recovery Plan Group Type. Possible values are `Boot`, `Failover` and `Shutdown`.

* `replicated_protected_items` - (required) one or more id of protected VM.

* `pre_action` - (Optional) one or more `action` block. which will be executed before the group recovery.

* `post_action` - (Optional) one or more `action` block. which will be executed after the group recovery.

---

An `action` block supports the following:

* `name` - (Required) Name of the Action.

* `action_detail_type` - (Required) Type of the action detail. Possible values are `AutomationRunbookActionDetails`, `ManualActionDetails` and `ScriptActionDetails`.

> * `fail_over_directions` - (Required) Directions of fail over. Must be one of `PrimaryToRecovery` or `RecoveryToPrimary`

Reviewer: ```suggestion
* `fail_over_directions` - (Required) Directions of fail over. Possible values are `PrimaryToRecovery` and `RecoveryToPrimary`
```

* `fail_over_types` - (Required) Types of fail over. Possible values are `TestFailover`, `PlannedFailover` and `UnplannedFailover`

> * `fabric_location` - (Optional) The fabric location of runbook or script. Must be one of `Primary` or `Recovery`.

Reviewer: ```suggestion
* `fabric_location` - (Optional) The fabric location of runbook or script. Possible values are `Primary` and `Recovery`.
```
